File: daily_problems/2025/07/0724/personal_submission/cf847f_yawn_sean.py
# Submission link: https://codeforces.com/contest/847/submission/330527172
def main():
    n, k, m, a = MII()
    nums = LII()

    ans = [2] * (n + 1)

    for i in range(1, n + 1):
        # 考虑最有利的情况
        
        cnt = [0] * (n + 1)
        last_vote = [-1] * (n + 1)
        
        for j in range(a):
            cnt[nums[j]] += 1
            last_vote[nums[j]] = j
        
        for j in range(a, m):
            cnt[i] += 1
            last_vote[i] = j
        
        if cnt[i] == 0:
            ans[i] = 3
            continue
        
        else:
            total = 1
            for j in range(1, n + 1):
                if cnt[j] > cnt[i] or (cnt[j] == cnt[i] and last_vote[j] < last_vote[i]):
                    total += 1
            if total > k:
                ans[i] = 3
                continue
        
        # 考虑最不利的情况
        
        cnt = [0] * (n + 1)
        last_vote = [-1] * (n + 1)
        
        for j in range(a):
            cnt[nums[j]] += 1
            last_vote[nums[j]] = j
        
        for tmstamp in range(a, m):
            
            # 选择紧随其后的对手
            cur_cnt, idx = -1, 0
            for j in range(1, n + 1):
                if cnt[j] < cnt[i] or (cnt[j] == cnt[i] and last_vote[j] > last_vote[i]):
                    if cnt[j] > cur_cnt:
                        cur_cnt = cnt[j]
                        idx = j
            
            cnt[idx] += 1
            last_vote[idx] = tmstamp

        if cnt[i]:
            total = 1
            for j in range(1, n + 1):
                if cnt[j] > cnt[i] or (cnt[j] == cnt[i] and last_vote[j] < last_vote[i]):
                    total += 1
            if total <= k:
                ans[i] = 1

    print(' '.join(map(str, ans[1:])))

File: daily_problems/2025/07/0724/personal_submission/cf847f_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, k, m, a;
    cin >> n >> k >> m >> a;

    vector<int> nums(a);
    for (auto &v: nums) cin >> v;

    for (int i = 1; i <= n; i ++) {
        {
            vector<int> cnt(n + 1, 0), last_vote(n + 1, -1);

            for (int j = 0; j < a; j ++) {
                cnt[nums[j]] ++;
                last_vote[nums[j]] = j;
            }

            for (int j = a; j < m; j ++) {
                cnt[i] ++;
                last_vote[i] = j;
            }

            if (cnt[i] == 0) {
                cout << 3 << ' ';
                continue;
            }
            else {
                int total = 1;
                for (int j = 1; j <= n; j ++) {
                    if (cnt[j] > cnt[i] || (cnt[j] == cnt[i] && last_vote[j] < last_vote[i])) {
                        total ++;
                    }
                }

                if (total > k) {
                    cout << 3 << ' ';
                    continue;
                }
            }
        }
        {
            vector<int> cnt(n + 1, 0), last_vote(n + 1, -1);

            for (int j = 0; j < a; j ++) {
                cnt[nums[j]] ++;
                last_vote[nums[j]] = j;
            }

            for (int tmstamp = a; tmstamp < m; tmstamp ++) {
                int cur_cnt = -1, idx = 0;
                for (int j = 1; j <= n; j ++) {
                    if (cnt[j] < cnt[i] || (cnt[j] == cnt[i] && last_vote[j] > last_vote[i])) {
                        if (cnt[j] > cur_cnt) {
                            cur_cnt = cnt[j];
                            idx = j;
                        }
                    }
                }
                cnt[idx] ++;
                last_vote[idx] = tmstamp;
            }

            if (cnt[i]) {
                int total = 1;
                for (int j = 1; j <= n; j ++) {
                    if (cnt[j] > cnt[i] || (cnt[j] == cnt[i] && last_vote[j] < last_vote[i])) {
                        total ++;
                    }
                }

                if (total <= k) {
                    cout << 1 << ' ';
                    continue;
                }
            }
        }
        cout << 2 << ' ';
    }

    return 0;
}